Added tree based folder navigational support for mods, and a way to install Fabric modpacks as folders.

abhicommands opened the pull request:
fixes #154 

Adds a full UI navigational system for folders in mods, with the .index folders and the metadata of mods in .pw.toml inside each subfolder of the respective mods. Mods can be downloaded into respective folders as well. A very highly requested feature that is really popular amongst users. 

Also adds a feature from issue #4560 Regarding modpacks being installed as folders instead of in the root to avoid any duplicates and easy control of updating and installing modpacks. 

Please do ask for any questions or concerns about this PR. 

Potential changes that I could add is a support to create new folders inside mods tab, and making the UI more snappy while expanding the folder. (need help with this one not able to do it in my own)

Ryex commented on launcher/minecraft/mod/tasks/ModFolderLoadTask.cpp:
```diff
@@ -0,0 +1,159 @@
+void ModFolderLoadTask::executeTask()
+{
+    if (thread() != m_thread_to_spawn_into)
+        connect(this, &Task::finished, this->thread(), &QThread::quit);
```
```suggestion
    if (thread() != m_thread_to_spawn_into) {
        connect(this, &Task::finished, this->thread(), &QThread::quit);
    }
```
Please always use braces for conditional or loop blocks. unless you putting them on one line like
```cpp
if (test) one_line_expression.op();
```
It's much easier to read
also, please run clang-format and clang-tidy. the code is wildly out of style with the rest of the codebase.

Ryex replied:
while the above lines are addressed the pattern still appears more than 10 times in other files in the commits

Ryex reviewed (commented):
Did a quick scan, honestly a bit confused about some of the design decisions but I'll take a closer look later

Ryex commented:
also, looks like the resource folder model tests are bork. pls fix that.

abhicommands commented:
Okay sure will look into it.

abhicommands commented:
> also, looks like the resource folder model tests are bork. pls fix that.

I think the first commit was working fine, not a lot of changes its a very simple change, that just allows users to add modpacks as folders instead of mods. The second one is obviously a very big change with sub folder support hence its taking a lot of time to understand and design properly, and I am not that well versed in building these types of app so will see how I can fix that. Could you just review the first commit to make sure that that one works? how can I make seperate requests for both those features?

abhicommands commented:
> Did a quick scan, honestly a bit confused about some of the design decisions but I'll take a closer look later

happy to clarify the design. The main goal here is to keep metadata colocated with the mods it describes (per‑folder .index), so subfolder installs behave consistently for updates, homepage links, and exports. The folder tree UI reflects the on‑disk layout directly, which makes selection/update paths deterministic and avoids guessing where a mod should be written back to.

So if mods are in a folder, then their metadata is stored in .index inside that specific subfolder, for mods in root mods folder it stays in root mods folder .index. Please let me know if you have further questions have fixed the tests that weren't working should build correctly now!

abhicommands commented:
@Ryex Have fixed the clang-tidy stuff and fixed some code quality as well as readability please let me know if they seem fine from ur end. Also if you could let the github workflows run to see if build works. Other than that everything works I think from my end. Please let me know any other work I need to finish up. And thank you @Trial97 for the added tags, but PLS can you make it 10.0.1 or something? I can look at dependency stuff too would love to have both those features in 10.0.1.

Trial97 commented on launcher/minecraft/mod/tasks/ModFolderLoadTask.cpp:
```diff
@@ -0,0 +1,192 @@
+// SPDX-License-Identifier: GPL-3.0-only
+/*
+ *  Prism Launcher - Minecraft Launcher
+ *  Copyright (c) 2024
```
add your name and email

Trial97 commented on launcher/minecraft/mod/tasks/ModFolderLoadTask.cpp:
```diff
@@ -0,0 +1,192 @@
+        return false;
+    }
+
+    QString name = info.fileName().toLower();
```
you can use info.suffix().toLower() saved in a variable and check that with equality instead of ends with

Trial97 replied:
and in this case I expect at least the disabled check to be done with the suffix equality.
Or at least instead of using the filename I would use `completeSuffix` function on QFileInfo

Trial97 commented on launcher/minecraft/mod/tasks/ModFolderLoadTask.cpp:
```diff
@@ -0,0 +1,192 @@
+    }
+
+    if (m_aborted) {
+        emit finished();
```
This line is bad: 
you should call emitAborted() function or at least have an additional call to ` emit aborted();` 
as we expect a task to finish in one of the 3 states: success,fail or abort 

Disclaimer I do understand that the `ResourceFolderLoadTask` has a similar issue: feel free to fix it there too but that is not necessary into this PR.

Disclaimer no2 I realize task  will never be aborted but still we should keep the same logic.

Trial97 commented on launcher/minecraft/mod/tasks/ModFolderLoadTask.h:
```diff
@@ -0,0 +1,66 @@
+// SPDX-License-Identifier: GPL-3.0-only
+/*
+ *  Prism Launcher - Minecraft Launcher
+ *  Copyright (c) 2024
```
add yorself please

Trial97 commented on launcher/modplatform/modrinth/ModrinthCheckUpdate.cpp:
```diff
@@ -171,7 +173,9 @@ void ModrinthCheckUpdate::checkVersionsResponse(std::shared_ptr<QByteArray> resp
             if ((project_ver.hash != hash && project_ver.is_preferred) || (resource->status() == ResourceStatus::NOT_INSTALLED)) {
-                auto download_task = makeShared<ResourceDownloadTask>(pack, project_ver, m_resourceModel);
+                QDir target_dir(resource->fileinfo().absolutePath());
+                auto download_task = makeShared<ResourceDownloadTask>(pack, project_ver, m_resourceModel, true, target_dir.absolutePath(),
```
this will install resources into the main folder and keep the old files into subdirectories

abhicommands replied:
The update target is based on the existing mod’s location, not the root.

Right above that call, target_dir is constructed from resource->fileinfo().absolutePath(). That means: If the mod is in mods/subfolder/, target_dir is that subfolder and the download goes there.
If it’s in mods/, it goes to root (as before).
ResourceDownloadTask also removes the old file using the same m_targetDir, so it won’t leave stale files behind in subfolders.

If there’s a spot where target_dir still points to root, please point me at it and I’ll fix it, but the current code should preserve subfolder updates.

Trial97 commented on launcher/InstanceImportTask.cpp:
This change is not needed for this PR.
Please don't touch more than necessary, as it can distract from the reviews.

Trial97 commented on launcher/ui/pages/instance/ModFolderPage.h:
```diff
@@ -59,23 +62,53 @@ class ModFolderPage : public ExternalResourcesPage {
+    QString relativeFolderPath(const QDir& dir) const;
+    QString normalizedModPath(const QFileInfo& file_info) const;
+=======
+>>>>>>> theirs
```
conflicts remains

Trial97 commented on launcher/ui/pages/instance/ModFolderPage.h:
```diff
@@ -59,23 +62,53 @@ class ModFolderPage : public ExternalResourcesPage {
+    void disableItem() override;
+    void viewHomepage() override;
+    bool eventFilter(QObject* obj, QEvent* ev) override;
+<<<<<<< ours
```
more conflict remains

Trial97 commented on launcher/ui/dialogs/ResourceUpdateDialog.cpp:
```diff
@@ -264,6 +264,7 @@ void ResourceUpdateDialog::checkCandidates()
 auto ResourceUpdateDialog::ensureMetadata() -> bool
 {
     auto index_dir = indexDir();
+    auto index_dir_resolver = [this](Resource* resource) { return m_resourceModel->indexDirForResource(*resource); };
```
Not necesary a required change but I would use std::bind here to not write this lambda or give the address directly

Trial97 commented on launcher/modplatform/modrinth/ModrinthInstanceCreationTask.h:
```diff
@@ -58,4 +58,7 @@ class ModrinthCreationTask final : public InstanceCreationTask {
     QString m_root_path = "minecraft";
+
+    bool m_use_mods_subdir = false;
+    QString m_mods_subdir_name;
```
```suggestion
    bool m_useModsSubdir = false;
    QString m_modsSubdir_name;
```

Trial97 commented on launcher/modplatform/modrinth/ModrinthInstanceCreationTask.cpp:
```diff
@@ -179,6 +242,33 @@ bool ModrinthCreationTask::createInstance()
     if (m_files.empty() && !parseManifest(index_path, m_files, true, true))
         return false;
 
+    const bool is_fabric_pack =
```
try using camelCase as much as posible

Trial97 commented on launcher/modplatform/flame/FlameCheckUpdate.cpp:
```diff
@@ -121,7 +123,9 @@ void FlameCheckUpdate::getLatestVersionCallback(Resource* resource, std::shared_
 
-        auto download_task = makeShared<ResourceDownloadTask>(pack, latest_ver.value(), m_resourceModel);
+        QDir target_dir(resource->fileinfo().absolutePath());
+        auto download_task = makeShared<ResourceDownloadTask>(pack, latest_ver.value(), m_resourceModel, true, target_dir.absolutePath(),
```
Same as modrinth update this will download to the main directory and keep the old files if they need update and are in subfolder

Trial97 reviewed (changes requested):
first round for this PR

abhicommands committed:
feat(mods): add folder tree support and subfolder modpack installs

- render mods in a folder tree (folders + mods)
- store mod metadata in per-folder .index directories
- keep updates/downloads in the mod’s existing folder
- fix resource folder model behavior and tests
- allow modpacks to install into mods/<pack-name> subfolders

Signed-off-by: abhicommands <114682464+abhicommands@users.noreply.github.com>

abhicommands commented:
Hey @Trial97 I have fixed all the things u asked and I kinda messed up the github commit history and sign off issues, hence it shows as only 1 commit, but its the same thing as before and now it consists of a cleaner commit history. 😂 But I think I resolved all of your comments, and what not, let me know if this works, also would love to see if the build and all the other workflow tests pass so if you could run that as well;

Ryex commented:
the single commit thing is a real problem 

please use

```
git reset --soft HEAD~1
```

then use

```
git add --patch
```
or
```
git add --interactive
```
or your editor/tool of choice to stage your changes step by step into a series of review able atomic commits.

preferably each commit containes a complete idea.

like:

  - add a setting for sub folders usage.
  - track subfolders with `.index`.
  - etc. etc.

Trial97 commented on launcher/minecraft/mod/tasks/ModFolderLoadTask.cpp:
```diff
@@ -0,0 +1,206 @@
+            QString chopped_id = resource->internal_id();
+            QFileInfo chopped_info(chopped_id);
+            if (chopped_info.suffix().compare("disabled", Qt::CaseInsensitive) == 0) {
+                chopped_id.chop(QString(".disabled").size());
```
```suggestion
                chopped_id.chop(9);
```

Trial97 commented on launcher/minecraft/mod/tasks/ModFolderLoadTask.cpp:
```diff
@@ -0,0 +1,206 @@
+{
+    auto rel_path = m_rootDir.relativeFilePath(info.absoluteFilePath());
+    rel_path = QDir::cleanPath(QDir::fromNativeSeparators(rel_path));
+    if (rel_path.startsWith("..")) {
```
are we expecting files outside the root directory?

abhicommands replied:
We don’t expect files outside m_rootDir. The .. guard is defensive (symlinks/odd paths) so we don’t store an escaping internal ID; in that case we fall back to fileName() to keep IDs stable. Let me know if you’d prefer we assert instead.

Trial97 commented on launcher/minecraft/mod/ModFolderTreeModel.h:
```diff
@@ -0,0 +1,93 @@
+// SPDX-License-Identifier: GPL-3.0-only
+/*
+ *  Prism Launcher - Minecraft Launcher
+ *  Copyright (c) 2024
```
add your name and email see other files for examples on how to

abhicommands replied:
08dbd27bb

Fixed

Trial97 commented on launcher/modplatform/flame/FlameCheckUpdate.cpp:
```diff
@@ -121,7 +123,9 @@ void FlameCheckUpdate::getLatestVersionCallback(Resource* resource, std::shared_
         }
 
-        auto download_task = makeShared<ResourceDownloadTask>(pack, latest_ver.value(), m_resourceModel);
+        QDir target_dir(resource->fileinfo().absolutePath());
```
Why do you need the conversion to QDir? can't you just use the string returned by the absolutePath function?
also convert snake_case to camelCase

abhicommands replied:
08dbd27bb

Trial97 commented on launcher/modplatform/flame/FlameCheckUpdate.cpp:
```diff
@@ -121,7 +123,9 @@ void FlameCheckUpdate::getLatestVersionCallback(Resource* resource, std::shared_
-        auto download_task = makeShared<ResourceDownloadTask>(pack, latest_ver.value(), m_resourceModel);
+        QDir target_dir(resource->fileinfo().absolutePath());
+        auto download_task = makeShared<ResourceDownloadTask>(pack, latest_ver.value(), m_resourceModel, true, target_dir.absolutePath(),
+                                                              target_dir.filePath(".index"), !resource->enabled());
```
Here you can use FS:CombinePath function to combine the targetDir string with .index

Trial97 commented on launcher/modplatform/modrinth/ModrinthCheckUpdate.cpp:
```diff
@@ -171,7 +173,9 @@ void ModrinthCheckUpdate::checkVersionsResponse(std::shared_ptr<QByteArray> resp
             pack->provider = ModPlatform::ResourceProvider::MODRINTH;
             if ((project_ver.hash != hash && project_ver.is_preferred) || (resource->status() == ResourceStatus::NOT_INSTALLED)) {
-                auto download_task = makeShared<ResourceDownloadTask>(pack, project_ver, m_resourceModel);
+                QDir target_dir(resource->fileinfo().absolutePath());
```
same as curseforge: do not convert to QDir and please use camelCase

Trial97 commented on launcher/modplatform/EnsureMetadataTask.cpp:
```diff
@@ -484,10 +495,13 @@ void EnsureMetadataTask::updateMetadata(ModPlatform::IndexedPack& pack, ModPlatf
-            ver.fileName.chop(9);
+        QFileInfo ver_info(ver.fileName);
+        if (ver_info.suffix().compare("disabled", Qt::CaseInsensitive) == 0) {
+            ver.fileName.chop(QString(".disabled").size());
```
Does the size of a constant change? Or why do we need to compute it?
```suggestion
            ver.fileName.chop(9);
```

abhicommands replied:
Edits made, sorry I didn't do it through Sign off and commit suggestion, but I fixed it manually

08dbd27bb

Trial97 commented on launcher/ui/dialogs/ResourceUpdateDialog.cpp:
```diff
@@ -264,6 +272,7 @@ void ResourceUpdateDialog::checkCandidates()
 auto ResourceUpdateDialog::ensureMetadata() -> bool
 {
     auto index_dir = indexDir();
+    auto index_dir_resolver = std::bind(&resolveIndexDir, m_resourceModel.get(), std::placeholders::_1);
```
I meant soomething like this:
```suggestion
    auto indexDirResolver =
    std::bind(&ResourceFolderModel::indexDirForResource,
              m_resourceModel.get(),
              std::placeholders::_1);
```
But if you think it is worse please go back to just using lambda instead of an external fucntion

Trial97 replied:
also camelCase is prefered

abhicommands replied:
08dbd27bb

Resolved, again sorry I didn't sign off and commit the suggestion.

Trial97 commented on launcher/BaseInstance.cpp:
```diff
@@ -122,6 +122,8 @@ BaseInstance::BaseInstance(SettingsObjectPtr globalSettings, SettingsObjectPtr s
     m_settings->registerSetting("ManagedPackName", "");
     m_settings->registerSetting("ManagedPackVersionID", "");
     m_settings->registerSetting("ManagedPackVersionName", "");
+    m_settings->registerSetting("ManagedPackUseModsSubdir", false);
+    m_settings->registerSetting("ManagedPackModsSubdirName", "");
```
Is this setting only for managed instances?
Isn't it for every type of instance?
The name of the setting should reflect that.

abhicommands replied:
These are only used by managed packs (Modrinth/Flame). Renaming would require a settings migration, so I kept the ManagedPack* prefix to match actual usage.

Trial97 commented on launcher/ResourceDownloadTask.cpp:
```diff
@@ -32,11 +34,24 @@
+        QFileInfo file_info(m_downloadFileName);
+        if (file_info.suffix().compare("disabled", Qt::CaseInsensitive) != 0) {
+            m_downloadFileName += ".disabled";
+        }
```
I'm not really against this but what is the reasoning of the user that updates the mod/resource?
Wouldn't it want to be enabled on update?

abhicommands replied:
Updates should respect the user’s explicit disabled state. m_keepDisabled preserves that choice so an update doesn’t unexpectedly re‑enable a mod.

Its a Design choice, and I think I might have instinctively fixed that, I think this is the correct way to do this, also I remember there was a issue and a pr about how disabled and updated mods were duplicated or something so, to avoid such cases as well.

Trial97 commented on launcher/ui/pages/instance/ModFolderPage.cpp:
```diff
@@ -317,30 +602,70 @@ void ModFolderPage::changeModVersion()
+    auto selection = m_filterModel->mapSelectionToSource(ui->treeView->selectionModel()->selection());
+    auto target_dir = m_treeModel->targetDirForSelection(selection.indexes());
+
+    QString name =
+        QInputDialog::getText(this, tr("New Folder"), tr("Enter a new folder name."), QLineEdit::Normal, tr("New Folder"), &ok).trimmed();
```
This is just a concern about the lifeline of this event in the big qt event loop as this line will block the thread.
This is raised as we used to have similar places where we displayed a dialog in an action and the parent was destroyed, causing some weird behaviour.

Again, this one is not a code change request but just a warning that may create some unwanted results.

abhicommands replied:
Thanks for the heads‑up. This is a modal dialog owned by the page instance; I’ll keep an eye on any lifecycle issues and can revisit if we see problems in practice.

Trial97 commented on launcher/ui/pages/instance/ModFolderPage.cpp:
```diff
@@ -317,30 +602,70 @@ void ModFolderPage::changeModVersion()
+        return;
+    }
+
+    if (name.isEmpty()) {
```
I would treat an empty as a cancel, not a new folder thing.

abhicommands replied:
Fixed. 

08dbd27bb

Trial97 commented on launcher/ui/pages/instance/ModFolderPage.cpp:
```diff
@@ -317,30 +602,70 @@ void ModFolderPage::changeModVersion()
+        name = tr("New Folder");
+    }
+
+    name = FS::RemoveInvalidFilenameChars(name, '-');
```
I would just call this before the isEmpty check minimizing the need of 2 isEmpty check

abhicommands replied:
fixed
 08dbd27bb

Trial97 commented on launcher/ui/pages/instance/ModFolderPage.cpp:
```diff
@@ -317,30 +602,70 @@ void ModFolderPage::changeModVersion()
+        name = tr("New Folder");
+    }
+
+    auto folder_name = FS::DirNameFromString(name, target_dir.absolutePath());
```
So how many times do you need to call RemoveInvalidFilenameChars? this one is already the second time.

Trial97 replied:
also please use camelCase

abhicommands replied:
Fixed

08dbd27bb

Trial97 commented on launcher/ui/pages/instance/ModFolderPage.cpp:
```diff
@@ -317,30 +602,70 @@ void ModFolderPage::changeModVersion()
+    }
+
+    auto folder_name = FS::DirNameFromString(name, target_dir.absolutePath());
+    if (folder_name.isEmpty()) {
```
explain how this can be empty?

Trial97 replied:
Like you have too many isEmpty checks in this function

abhicommands replied:
can return empty when it can’t generate a valid/unique folder name in the target dir. I kept a single guard for that failure; other redundant isEmpty() checks were removed.”

Fixed: 

08dbd27bb

Trial97 commented on launcher/ui/pages/instance/ModFolderPage.cpp:
```diff
@@ -317,30 +602,70 @@ void ModFolderPage::changeModVersion()
-    std::sort(selectedMods.begin(), selectedMods.end(), [](const Mod* a, const Mod* b) { return a->name() < b->name(); });
-    ExportToModListDialog dlg(m_instance->name(), selectedMods, this);
+    auto mods = m_model->allMods();
+    std::sort(mods.begin(), mods.end(), [](const Mod* a, const Mod* b) { return a->name() < b->name(); });
```
revert this: I would still like the user to be able to select what it wants to export

abhicommands replied:
Fixed:
08dbd27bb

Trial97 commented on launcher/ui/pages/instance/ModFolderPage.cpp:
```diff
@@ -133,16 +302,119 @@ void ModFolderPage::removeItems(const QItemSelection& selection)
+{
+    auto selection = m_filterModel->mapSelectionToSource(ui->treeView->selectionModel()->selection()).indexes();
+    auto resources = m_treeModel->resourcesFromIndexes(selection);
+    auto indexes = m_model->indexesForResources(resources);
```
This lines repeats more than 3 time: make it a function

abhicommands replied:
Fixed: 
08dbd27bb

Trial97 commented on launcher/ui/pages/instance/ModFolderPage.cpp:
```diff
@@ -133,16 +302,119 @@ void ModFolderPage::removeItems(const QItemSelection& selection)
+        }
+        auto dir = m_treeModel->folderForIndex(index);
+        auto path = QDir::cleanPath(dir.absolutePath());
+        if (path == QDir::cleanPath(m_model->dir().absolutePath())) {
```
You do not need to acll clean path on absolute path

abhicommands replied:
fixed

08dbd27bb

Trial97 commented on launcher/ui/pages/instance/ModFolderPage.cpp:
```diff
@@ -133,16 +302,119 @@ void ModFolderPage::removeItems(const QItemSelection& selection)
+    QStringList pruned_folders = folder_paths.values();
+    std::sort(pruned_folders.begin(), pruned_folders.end(),
+              [](const QString& left, const QString& right) { return left.size() < right.size(); });
+    QStringList folders_to_remove;
```
camelCase

abhicommands replied:
fixed: 

08dbd27bb

Trial97 commented on launcher/ui/pages/instance/ModFolderPage.cpp:
```diff
@@ -133,16 +302,119 @@ void ModFolderPage::removeItems(const QItemSelection& selection)
+
+    for (const auto& folder : folders_to_remove) {
+        if (!FS::deletePath(folder)) {
+            CustomMessageBox::selectable(this, tr("Error"), tr("Failed to remove folder '%1'.").arg(folder), QMessageBox::Warning)->show();
```
bad design: only show the error message once, not for each folder it fails to delete.
Otherwise, you will get an annoying amount of messages shown to you for each folder(if you have 100 folders that fail, you get 100 selectable, which is not  nice)

abhicommands replied:
Fixed 
08dbd27bb

Trial97 commented on launcher/ui/pages/instance/ModFolderPage.cpp:
```diff
@@ -133,16 +302,119 @@ void ModFolderPage::removeItems(const QItemSelection& selection)
+        }
+    }
+
+    if (!folders_to_remove.isEmpty()) {
```
Shouldn't we always update the model?
You do not delete from the list, so this can be moved before the last folder and exit early from the function. 
But again, the question remains: shouldn't the model be updated regardless of whether we delete folders or not?

abhicommands replied:
Fixed
08dbd27bb

Trial97 commented on launcher/ui/pages/instance/ModFolderPage.cpp:
```diff
@@ -133,16 +302,119 @@ void ModFolderPage::removeItems(const QItemSelection& selection)
+        if (index.column() != 0) {
+            continue;
+        }
+        if (!m_treeModel->isFolderIndex(index)) {
```
Do we display the .index folder to the user? Or what is this check for?

abhicommands replied:
We don’t display .index in the tree it’s internal metadata. The check skips it so it can’t be selected/removed as a folder

Trial97 commented on launcher/ui/pages/instance/ModFolderPage.cpp:
```diff
@@ -109,19 +121,176 @@ ModFolderPage::ModFolderPage(BaseInstance* inst, std::shared_ptr<ModFolderModel>
+        if (!is_child) {
+            folders_to_remove.append(path);
+        }
+    }
```
This can be made as a fucntion

abhicommands replied:
Fixed

08dbd27bb

Trial97 commented on launcher/ui/pages/instance/ModFolderPage.cpp:
```diff
@@ -133,16 +302,119 @@ void ModFolderPage::removeItems(const QItemSelection& selection)
+
+    QStringList pruned_folders = folder_paths.values();
+    std::sort(pruned_folders.begin(), pruned_folders.end(),
+              [](const QString& left, const QString& right) { return left.size() < right.size(); });
```
sort after qstring comparation not only the lenght

abhicommands replied:
Fixed:

08dbd27bb

Trial97 commented on launcher/ui/pages/instance/ModFolderPage.cpp:
```diff
@@ -109,19 +121,176 @@ ModFolderPage::ModFolderPage(BaseInstance* inst, std::shared_ptr<ModFolderModel>
+    auto selection_indexes = selection.indexes();
+    auto resources = m_treeModel->resourcesFromIndexes(selection_indexes);
+    QSet<QString> folder_paths;
+    for (const auto& index : selection_indexes) {
```
so we expect only one item but still iterate over all?

abhicommands replied:
Selection can include multiple rows. We iterate to collect all selected folders/resources. This is intentional and still supports single‑selection.

Trial97 commented on launcher/ui/pages/instance/ModFolderPage.cpp:
```diff
@@ -109,19 +121,176 @@ ModFolderPage::ModFolderPage(BaseInstance* inst, std::shared_ptr<ModFolderModel>
+    if (!folders_to_remove.isEmpty() || resources.size() > 1 || (resources.size() == 1 && !folders_to_remove.isEmpty())) {
+        QString message;
+        if (folders_to_remove.size() == 1 && resources.isEmpty()) {
+            message = tr("You are about to remove the folder '%1'.\n"
```
Nice we have confirmation when we delete one folder but not when deleting multiple: so let's annoy the user to confirm the folder deletion only when he explicitly selects one?

abhicommands replied:
Fixed

08dbd27bb

Thanks for pointing that out, I’ve updated the confirmation flow so it applies consistently for single and multiple folder deletions.

Trial97 commented on launcher/ui/pages/instance/ModFolderPage.cpp:
```diff
@@ -109,19 +121,176 @@ ModFolderPage::ModFolderPage(BaseInstance* inst, std::shared_ptr<ModFolderModel>
+    ui->frame->updateWithMod(*mod);
+}
+
+void ModFolderPage::removeItem()
```
I like how the name of the function is misleading. This is just a wrapper around `removeItems` that also displays some confirmation for the user and also does some of the `removeItems` work again.

abhicommands replied:
08dbd27bb

Resolved

Trial97 commented on launcher/ui/pages/instance/ModFolderPage.cpp:
```diff
@@ -109,19 +121,176 @@ ModFolderPage::ModFolderPage(BaseInstance* inst, std::shared_ptr<ModFolderModel>
+void ModFolderPage::updateActions()
+{
+    const auto selection = m_filterModel->mapSelectionToSource(ui->treeView->selectionModel()->selection()).indexes();
+    const auto selected_resources = m_treeModel->resourcesFromIndexes(selection);
```
camelCase

abhicommands replied:
08dbd27bb

Trial97 commented on launcher/ui/pages/instance/ModFolderPage.cpp:
```diff
@@ -109,19 +121,176 @@ ModFolderPage::ModFolderPage(BaseInstance* inst, std::shared_ptr<ModFolderModel>
+    const auto selected_resources = m_treeModel->resourcesFromIndexes(selection);
+
+    const bool has_mod_selection = !selected_resources.isEmpty();
+    bool has_folder_selection = false;
```
camelCase

abhicommands replied:
08dbd27bb

Trial97 commented on launcher/ui/pages/instance/ModFolderPage.cpp:
```diff
@@ -109,19 +121,176 @@ ModFolderPage::ModFolderPage(BaseInstance* inst, std::shared_ptr<ModFolderModel>
+        }
+    };
+    connect(selection_model, &QItemSelectionModel::selectionChanged, this, updateExtra);
+    connect(selection_model, &QItemSelectionModel::selectionChanged, this, [this] { updateActions(); });
```
```suggestion
    connect(selection_model, &QItemSelectionModel::selectionChanged, this, &ModFolderPage::updateActions);
```

Trial97 commented on launcher/ui/pages/instance/ModFolderPage.cpp:
```diff
@@ -109,19 +121,176 @@ ModFolderPage::ModFolderPage(BaseInstance* inst, std::shared_ptr<ModFolderModel>
+
+    auto selection_model = ui->treeView->selectionModel();
+    connect(selection_model, &QItemSelectionModel::currentChanged, this, [this](const QModelIndex& current, const QModelIndex& previous) {
+        if (!current.isValid()) {
```
Just move this check in updateFrame function and connect it directly to it and do not use a lambda.

abhicommands replied:
08dbd27bb

Trial97 reviewed (changes requested):
Also please note that on new files that you add I still want the copyright to include your name and email(it doesn't need to be your real name, it can be the GitHub username)

abhicommands commented:
> the single commit thing is a real problem
> 
> please use
> 
> ```
> git reset --soft HEAD~1
> ```
> 
> then use
> 
> ```
> git add --patch
> ```
> 
> or
> 
> ```
> git add --interactive
> ```
> 
> or your editor/tool of choice to stage your changes step by step into a series of review able atomic commits.
> 
> preferably each commit containes a complete idea.
> 
> like:
> 
> * add a setting for sub folders usage.
> * track subfolders with `.index`.
> * etc. etc.

Hey This was just a slight mistake from my part this happened because, i wanted to refactor my commits, where I made an error and put my real email for dco sign off, I have fixed that now, and it won't happen again, so its like this I will fix this when I review all the changes previously we had 5 commits, with my real email so while fixing the history became a little wierd will be fixing all of that. Again sorry about that, i just didn't want that stuff to be on the dco sign off!

abhicommands committed:
Update launcher/minecraft/mod/tasks/ModFolderLoadTask.cpp

Co-authored-by: Alexandru Ionut Tripon <alexandru.tripon97@gmail.com>
Signed-off-by: Abhinav Acharya <114682464+abhicommands@users.noreply.github.com>

abhicommands committed:
Update launcher/ui/pages/instance/ModFolderPage.cpp

Co-authored-by: Alexandru Ionut Tripon <alexandru.tripon97@gmail.com>
Signed-off-by: Abhinav Acharya <114682464+abhicommands@users.noreply.github.com>

TheKodeToad commented:
> Please always use braces for conditional or loop blocks. unless you putting them on one line like

since this hasn't been fully addressed -
FWIW nobody ever told me to do any different to this and I was only trying to adhere to existing formatting when I started doing this years ago
If we wanna have a consistent style that's great but I'd like to see a PR enforcing it :)

abhicommands committed:
fix(ModFolder): Address Review Comments

Signed-off-by: abhicommands <114682464+abhicommands@users.noreply.github.com>

abhicommands committed:
Fixed my name and year in the license headers for the files.

Signed-off-by: abhicommands <114682464+abhicommands@users.noreply.github.com>

abhicommands commented:
@Trial97 I looked at all the variable names, and changed it to camel case, also started looking at clang-format and clang-tidy made sure the coding style is consistent, also fixed all the comments and resolved them. I also added my name on all the new files that were added with proper consistent formatting to other files as well. Please let me know if there are anything else I need to fix add etc, Thank you!

abhicommands committed:
Merge remote-tracking branch 'upstream/develop' into develop

abhicommands committed:
Merge branch 'PrismLauncher:develop' into develop

abhicommands committed:
Merge branch 'PrismLauncher:develop' into develop

abhicommands committed:
Merge branch 'PrismLauncher:develop' into develop

abhicommands committed:
Merge branch 'PrismLauncher:develop' into develop

abhicommands committed:
Merge branch 'PrismLauncher:develop' into develop

abhicommands committed:
fix(mods): refine foldered modpacks update flow

- add per-update subfolder toggle and JVM arg note
- scope update checks to root mods unless folders/mods are selected
- keep mod updates in their original folders
- improve tree model helpers used by update targeting

Signed-off-by: abhicommands <114682464+abhicommands@users.noreply.github.com>

abhicommands committed:
fix(formatting): Correct formatting for all modified files to adhere to project standards.

Signed-off-by: abhicommands <114682464+abhicommands@users.noreply.github.com>

abhicommands committed:
fix(formatting): Fixed some formatting issues raised by the clang-tidy, such as Braces etc. Other fixes are still pending

Signed-off-by: abhicommands <114682464+abhicommands@users.noreply.github.com>

abhicommands committed:
fix(formatting): fixed all the edits according to clang-tidy suggestions more edits impending!

Signed-off-by: abhicommands <114682464+abhicommands@users.noreply.github.com>

abhicommands committed:
formatting fix: corrected code to clang-tidy formatting standards.

Signed-off-by: abhicommands <114682464+abhicommands@users.noreply.github.com>

abhicommands committed:
Merge remote-tracking branch 'upstream/develop' into develop

abhicommands committed:
Merge branch 'PrismLauncher:develop' into develop

abhicommands committed:
formatting: fix indentation in ResourceDownloadTask.cpp

Signed-off-by: abhicommands <114682464+abhicommands@users.noreply.github.com>

abhicommands committed:
Merge branch 'PrismLauncher:develop' into develop

abhicommands committed:
Merge branch 'PrismLauncher:develop' into develop

abhicommands commented:
Closing this PR as superseded by #4988.
Moved the work to a new branch/PR with the virtual metadata-group architecture, managed-pack-safe update behavior, grouped update review UX, migration coverage, and test/build verification.

Please continue review on #4988.
